apps/src/code-studio/signup.js

```diff
@@ -0,0 +1,153 @@
+import $ from 'jquery';
+
+window.SignupManager = function (options) {
+  this.options = options;
+  var self = this;
+
+  // Check for URL having: /users/sign_up?user%5Buser_type%5D=teacher
+  if (self.options.isTeacher === "true") {
+    // Select teacher in dropdown.
+    $("#user_user_type").val("teacher");
+
+    showTeacher();
+  }
+
+  function formSuccess(success) {
+    var url;
+    if (self.options.returnToUrl !== "") {
+      url = self.options.returnToUrl;
+    } else if (isTeacherSelected()) {
+      url = self.options.teacherDashboardUrl;
+    } else {
+      url = "/";
+    }
+    window.location.href = url;
+  }
+
+  function formError(err) {
+    // Define the fields that can have specific errors attached to them.
+    var fields = ["user_type", "name", "email", "password", "password_confirmation", "schoolname", "age", "gender"];
+
+    for (var i = 0; i < fields.length; i++) {
+      var field = fields[i];
+      if (err.responseJSON.errors[field]) {
+        var errorField = $(`#${field}-block .error_in_field`);
+         // We have a custom inline message for user_type errors already set in the DOM.
+        if (field !== "user_type") {
+          errorField.text(err.responseJSON.errors[field][0]);
+        }
+        errorField.fadeTo("normal", 1);
+      }
+    }
+  }
+
+  $("#user_user_type").change(function () {
+    var value = $(this).val();
+    if (value === "student") {
+      showStudent();
+    } else if (value === "teacher") {
+      showTeacher();
+    }
+  });
+
+  function showStudent() {
+    // Show correct form elements.
+    $("#age-block").fadeIn();
+    $("#gender-block").fadeIn();
+    $("#schoolname-block").hide();
+
+    // Show correct terms below form.
+    $("#teacher-terms").hide();
+    $("#student-terms").fadeIn();
+  }
+
+  function showTeacher() {
+    // Show correct form elements.
+    $("#age-block").hide();
+    $("#gender-block").hide();
+    $("#schoolname-block").fadeIn();
+
+    // Show correct terms below form.
+    $("#student-terms").hide();
+    $("#teacher-terms").fadeIn();
+  }
+
+  function isTeacherSelected() {
+    var formData = $('#new_user').serializeArray();
+    var userType = $.grep(formData, e => e.name === "user[user_type]");
+    if (userType.length === 1 && userType[0].value === "teacher") {
+      return true;
+    }
+    return false;
+  }
+
+  $(".signupform").submit(function () {
+    // Clear the prior hashed email.
+    $('#user_hashed_email').val('');
+
+    // Hash the email.
+    window.dashboard.hashEmail({email_selector: "#user_email",
+      hashed_email_selector: "#user_hashed_email",
+      age_selector: "#user_user_age",
+      skip_clear_email: true});
+
+    var formData = $("#new_user").serializeArray();
+
+    if (isTeacherSelected()) {
+      // Teachers get age 21 in the form data.
+      formData.push({name: "user[age]", value: 21});
```

Comment:
Good catch.  Shouldn't affect what gets written to db in User model, but will fix this on client.